src/iop/lut3d.c

```diff
@@ -0,0 +1,1200 @@
+    else Norm = (lin[ch] - f) / e;
+    lout[ch] = Norm < 0.0f ? 0.0f : Norm > 1.0f ? 1.0f : Norm;
+  }
+}
+
+// From data/kernels/extended.cl
+static inline float fastlog2(float x)
+{
+  union { float f; unsigned int i; } vx = { x };
+  union { unsigned int i; float f; } mx = { (vx.i & 0x007FFFFF) | 0x3f000000 };
+
+  float y = vx.i;
+
+  y *= 1.1920928955078125e-7f;
+
+  return y - 124.22551499f
+    - 1.498030302f * mx.f
+    - 1.72587999f / (0.3520887068f + mx.f);
+}
+
+static inline void lin_to_log2(float *lin, float *lout, const float middle_grey,
+                                    const float min_exposure, const float dynamic_range)
+{
+  for (int ch=0; ch<3; ch++)
+  {
+    const float logNorm = (fastlog2(lin[ch] / middle_grey) - min_exposure) / dynamic_range;
+    if( logNorm < 0.0f || lin[ch] <= 0.0f)
+    {
+      lout[ch] = 0.0f;
+    }
+    else
+    {
+      lout[ch] = logNorm;
+    }
+  }
+}
+
+static inline void log2_to_lin(float *lin, float *lout, const float middle_grey,
+                                    const float min_exposure, const float dynamic_range)
+{
+  for (int ch=0; ch<3; ch++)
+  {
+    const float Norm = middle_grey * powf(2, lin[ch] * dynamic_range + min_exposure);
+    lout[ch] = Norm < 0.0f ? 0.0f : Norm > 1.0f ? 1.0f : Norm;
+  }
+}
+
+void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const void *const ibuf, void *const obuf,
+             const dt_iop_roi_t *const roi_in, const dt_iop_roi_t *const roi_out)
+{
+//printf("process\n");
+  dt_iop_lut3d_params_t *p = (dt_iop_lut3d_params_t *)self->params;
+  dt_iop_lut3d_global_data_t *gp = (dt_iop_lut3d_global_data_t *)self->data;
+  const int ch = piece->colors;
+  const int colorspace = p->colorspace;
+  const float middle_grey = p->middle_grey * 0.01f;
+  const float min_exposure = p->min_exposure;
+  const float dynamic_range = p->dynamic_range;
+  const int log2tolin = p->log2tolin;
+  const float *const clut = (float *)gp->clut;
+  const uint8_t level = gp->level;
+printf("process colorspace %d ch %d\n", colorspace, ch);
+  const dt_interpolation_worker interpolation_worker = (p->interpolation == DT_IOP_TETRAHEDRAL) ? correct_pixel_tetrahedral
+                                                          : (p->interpolation == DT_IOP_TRILINEAR) ? correct_pixel_trilinear
+                                                          : correct_pixel_pyramid;
+
+  const dt_iop_order_iccprofile_info_t *const dest_profile
+      = colorspace == DT_IOP_REC709 ? dt_ioppr_add_profile_info_to_list(self->dev, DT_COLORSPACE_REC709, "", INTENT_PERCEPTUAL)
+      : colorspace == DT_IOP_SRGB ? dt_ioppr_add_profile_info_to_list(self->dev, DT_COLORSPACE_SRGB, "", INTENT_PERCEPTUAL)
+      : colorspace == DT_IOP_LIN_REC709 ? dt_ioppr_add_profile_info_to_list(self->dev, DT_COLORSPACE_LIN_REC709, "", INTENT_PERCEPTUAL)
+      : NULL;
+  const dt_iop_order_iccprofile_info_t *const rgb_profile = dt_ioppr_get_iop_work_profile_info(self->dev);

```

Comment:
The work profile can be NULL, I know this is a work in progress, just so you know it.

Reply:
If work profile is NULL, does that mean the module is not between colorin and colorout ? If not what does that mean ?

Reply:
Yes, that's what it means.

Reply:
Should be dt_ioppr_get_pipe_work_profile_info(piece->pipe) here instead of dt_ioppr_get_iop_work_profile_info(self->dev)

Reply:
If work profile is NULL, how can I know if the module is before colorin or after colorout ?